default/python/AI/pathfinding.py

```diff
@@ -1,16 +1,137 @@
+import FreeOrionAI as foAI
+import freeOrionAIInterface as fo
+import universe_object
+
 from heapq import heappush, heappop
 from collections import namedtuple
 
-import freeOrionAIInterface as fo
 from AIDependencies import INVALID_ID
+from EnumsAI import MissionType
 from turn_state import state
-
 from common.configure_logging import convenience_function_references_for_logger
-(debug, info, warn, error, fatal) = convenience_function_references_for_logger(__name__)
+from freeorion_tools import cache_by_session_with_turnwise_update, chat_human, get_partial_visibility_turn
 
 
+(debug, info, warn, error, fatal) = convenience_function_references_for_logger(__name__)
+_DEBUG_CHAT = False
 _ACCEPTABLE_DETOUR_LENGTH = 2000
 
+# cache this so that boost python does not need to make a new copy every time this info is needed in a turn.
+@cache_by_session_with_turnwise_update
+def _get_unobstructed_systems():
+    return fo.getEmpire().supplyUnobstructedSystems
+
+def _info_string(path_info):
+    sequence_string = ", ".join([str(universe_object.System(sys_id)) for sys_id in path_info.path])
+    return "dist %.1f, path %s" % (path_info.distance, sequence_string)
+
+# Note that this can cover departure from uncontested systems as well as from contested systems where our forces
+# arrived first (and those forces may or may not be the forces for whom the pathfinding is being done)
+def _more_careful_travel_starlane_func(c, d):
+    return c in _get_unobstructed_systems()
+
+
+# In comparison to  _more_careful_travel_starlane_func, this can also allow passage through contested
+# systems where our forces already present there (which may or may not be the forces for whome the pathfinding
+# is being done) arrived second but are currently preserving travel along the starlane of interest.
+def _somewhat_careful_travel_starlane_func(c, d):
+    return any((c in _get_unobstructed_systems(),
```

Comment:
I had it like that for a bit, but it did not really seem worth the overhead of an extra python function call just for emphasizing that relationship.  I'm close to a tossup on that, though, so if you feel strongly about it it's fine with me, at least later once  the UnrestrictedLaneTravel() issue gets settled.  Depending on which resolution we go with for the UnrestrictedLaneTravel() issue, this whole reference to unobstructed systems might disappear because it would be subsumed in the unrestrictedLaneTravel check, so I'd prefer to leave it as-is for now.

Reply:
> the overhead of an extra python function call

Not a very strong argument if you use ```any``` to replace a single ```or``` in the very same line

I think it would be cleaner code because of the comment which states ```in comparison with _more_careful_travel_starlane_func``` which implies the relationship should hold regardless of the implementation. If you think you are going to change that in the not-too-far future, then I am fine with leaving it as is.

Reply:
Touchee.  I see that as more of a reason to change the ```any``` to an ```or```, I think I have gotten a bit carried away with ```any``` lately.  Anyways, let's wait until the UnrestrictedLaneTravel thing is resolved.  I am heavily leaning towards the solution there that is just clarifying the comments, plus changing the name as I mentioned, to be ```PreservedLaneTravel()```  -- if that makes sense to you it would be great to get some feedback on the Issue/PRs.